Задание 4.3. Напишите функцию brackets(line), которая определяет, является ли последовательность из круглых скобок правильной. Используйте структуру данных типа стек:
- Если встречается открывающая скобка, добавьте её в стек.
- Если встречается закрывающая скобка, удалите верхний элемент из стека.
- Если при удалении верхнего элемента стека он оказался пуст, то строка неверна.
- Если все элементы перебраны, а стек пуст то строка верна.
- Если все элементы перебраны, а стек не пуст, то строка неверна.

In [1]:
from collections import deque

In [3]:
def brackets(line):
    brackets = deque()

    for bracket in line:
        if bracket == "(":
            brackets.append(bracket)
        elif bracket == ")":
            if len(brackets) == 0:
                return False

            brackets.pop()

    return len(brackets) == 0

print(brackets("(()())"))
# True
print(brackets(""))
# True
print(brackets("(()()))"))
# False

True
True
False


### Задание 4.4

В переменных center, south и north хранятся списки из перечней купленных позиций в трёх торговых точках, расположенных в разных районах города.
Вначале избавьтесь от излишней вложенности: в каждой переменной (center, south, north) должен храниться объединённый список купленных товаров без разбиения по чекам.

Пример: [['Milk', 'Bread'], ['Meat']] -> ['Milk', 'Bread', 'Meat']

После этого определите, в каком магазине было куплено больше всего товаров.

In [4]:
from hidden import north, center, south

In [6]:
def serialize(items):
    serialized = []
    for row in items:
        serialized.extend(iter(row))

    return serialized


In [24]:
from hidden import north, center, south

from collections import Counter


north_counter = Counter(serialize(north))
north_total = sum(north_counter.values())
center_counter = Counter(serialize(center))
center_total = sum(center_counter.values())
south_counter = Counter(serialize(south))
south_total = sum(south_counter.values())

print(f"В магазине north куплено {north_total} товаров")
print(f"В магазине center куплено {center_total} товаров")
print(f"В магазине south куплено {south_total} товаров")

В магазине north куплено 500 товаров
В магазине center куплено 548 товаров
В магазине south куплено 478 товаров


### Задание 4.5

Теперь получите объекты-счётчики (Counter) из каждого полученного в предыдущем задании списка покупок и сохраните их в отдельные переменные (они пригодятся для выполнения следующих задач).
Сколько раз покупали самый редкий товар в магазине north? Запишите ответ в числовой форме.

In [25]:
print(min(north_counter.values()))

10


### Задание 4.6

Выберите товар, который в магазине center покупали чаще, чем в магазине north:

In [26]:
center_counter - north_counter

Counter({'Meat': 31,
         'Cheese': 29,
         'Milk': 21,
         'Bread': 23,
         'Chocolate': 2,
         'Chips': 22,
         'Ketchup': 19})

Bread присутствует в вариантах ответов и счетчике одновременно.  Ответ: Bread

### Задание 4.7

Есть ли такой товар, который в одном из магазинов покупали чаще, чем в двух других вместе взятых? Если да, выберите магазин с настолько популярным товаром:

In [29]:
(center_counter - north_counter) - south_counter

Counter()

In [30]:
(north_counter - center_counter) - south_counter

Counter({'Cola': 6})

Ответ: north

### Задание 4.8.

Определите суммарное число продаж каждого товара во всех магазинах, сложив все объекты-счётчики. Сколько раз был куплен второй по популярности товар? Запишите ответ в числовой форме.


In [33]:
total_counter = north_counter + center_counter + south_counter

sorted(total_counter.values(), reverse=True)[1]

240

Ответ: 240

### Задание 4.9

Дан список кортежей ratings с рейтингами кафе. Кортеж состоит из названия и рейтинга кафе.

Необходимо отсортировать список кортежей по убыванию рейтинга. Если рейтинги совпадают, то отсортировать кафе дополнительно по названию в алфавитном порядке.

Получите словарь cafes с упорядоченными ключами из отсортированного списка, где ключи — названия кафе, а значения — их рейтинг.

In [38]:
ratings = [
    ("Old York", 3.3),
    ("New Age", 4.6),
    ("Old Gold", 3.3),
    ("General Foods", 4.8),
    ("Belissimo", 4.5),
    ("CakeAndCoffee", 4.2),
    ("CakeOClock", 4.2),
    ("CakeTime", 4.1),
    ("WokToWork", 4.9),
    ("WokAndRice", 4.9),
    ("Old Wine Cellar", 3.3),
    ("Nice Cakes", 3.9),
]

# Отсортируйте список ratings по убыванию рейтинга. Для кафе
# с одинаковым рейтингом отсортируйте кортежи по названию.

ratings = sorted(ratings, key=lambda x: (-x[1], x[0]))

# Сохраните данные с рейтингом в словарь cafes, где ключами являются
# названия кафе, а значениями - их рейтинг.
from collections import OrderedDict

cafes = OrderedDict(ratings)
print(cafes)

OrderedDict([('WokAndRice', 4.9), ('WokToWork', 4.9), ('General Foods', 4.8), ('New Age', 4.6), ('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1), ('Nice Cakes', 3.9), ('Old Gold', 3.3), ('Old Wine Cellar', 3.3), ('Old York', 3.3)])


### Задание 4.10

Напишите функцию task_manager, которая принимает список задач для нескольких серверов. Каждый элемент списка состоит из кортежа (<номер задачи>, <название сервера>, <высокий приоритет задачи>).

Функция должна создавать словарь и заполнять его задачами по следующему принципу: название сервера — ключ, по которому хранится очередь задач для конкретного сервера. Если поступает задача без высокого приоритета (последний элемент кортежа — False), добавить номер задачи в конец очереди. Если приоритет высокий, добавить номер в начало.

Для словаря используйте defaultdict, для очереди — deque.

Функция возвращает полученный словарь с задачами.

Пример:
``` python
tasks = [(36871, 'office', False),
(40690, 'office', False),
(35364, 'voltage', False),
(41667, 'voltage', True),
(33850, 'office', False)]
 
print(task_manager(tasks))
# defaultdict(, {'voltage': deque([41667, 35364]),
# 'office': deque([36871, 40690, 33850])})
```

In [39]:
tasks = [(36871, 'office', False),
(40690, 'office', False),
(35364, 'voltage', False),
(41667, 'voltage', True),
(33850, 'office', False)]

In [49]:
from collections import defaultdict, deque


def task_manager(tasks):
    tasks_dict = defaultdict(deque)

    for task, server, priority in tasks:
        if priority:
            tasks_dict[server].appendleft(task)
        else:
            tasks_dict[server].append(task)
    
    return tasks_dict

In [50]:
task_manager(tasks)

defaultdict(collections.deque,
            {'office': deque([36871, 40690, 33850]),
             'voltage': deque([41667, 35364])})